common/djangoapps/util/milestones_helpers.py

```diff
@@ -0,0 +1,151 @@
+    if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES'):
+        for course, __ in enrolled_courses:
+            required_courses = []
+            # if course has pre-requisites then fetch fulfilment path
+            if course.pre_requisite_courses:
+                fulfilment_paths = get_course_milestones_fulfillment_paths(course.id, {'id': user.id})
+                for milestone_key, milestone_value in fulfilment_paths.items():  # pylint: disable=unused-variable
+                    for key, value in milestone_value.items():
+                        if key == 'courses' and value:
+                            for required_course in value:
+                                required_course_key = CourseKey.from_string(required_course['course_id'])
+                                required_course_descriptor = modulestore().get_course(required_course_key)
+                                required_courses.append({
+                                    'key': required_course_key,
+                                    'display': ' '.join([
+                                        required_course_descriptor.display_org_with_default,
+                                        required_course_descriptor.display_number_with_default
+                                    ])
+                                })
+
+                # if there are required courses add to dict
+            if required_courses:
+                pre_requisite_courses[course.id] = {'courses': required_courses}
+    return pre_requisite_courses
+
+
+def get_prerequisite_courses_display(course_descriptor):  # pylint: disable=invalid-name
+    """
+    It would retrieve pre-requisite courses, make display strings
+    and return them as list
+    """
+    pre_requisite_courses = []
+    if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES', False) and course_descriptor.pre_requisite_courses:
+        for course_id in course_descriptor.pre_requisite_courses:
+            course_key = CourseKey.from_string(course_id)
+            required_course_descriptor = modulestore().get_course(course_key)
+            pre_requisite_courses.append(' '.join([
+                required_course_descriptor.display_org_with_default,
+                required_course_descriptor.display_number_with_default
+            ]))
+    return pre_requisite_courses
+
+
+def fulfill_course_milestone(course_key, user):
+    """
+    It would save course milestone collected by user.
+    """
+    course_milestones = get_course_milestones(course_key=course_key, relationship="fulfills")
+    for milestone in course_milestones:
+        add_user_milestone({'id': user.id}, milestone)
+
+
+def milestones_achieved_by_user(user):
+    """
+    It would fetch list of milestones completed by user
+    """
+    return get_user_milestones({'id': user.id})
+
+
+def is_valid_course_key(key):
+    """
+    validates course key. returns True if valid else False.
+    """
+    try:
+        course_key = CourseKey.from_string(key)
```

Comment:
@dmitchell is it necessary to go through all of this to validate a course key (is there an existing validation method)? I thought CourseKey.from_string would handle route to SlashSeparatedCourseKey.from_deprecated_string if necessary.